### PR TITLE
Adds the `dojo create app` command

### PR DESCRIPTION
**Pull request by tomdye**
## Overview

Allows an app to be scaffolded by running `dojo create app` or `dojo create` if loaded as the only `create` command.
## Sample usage

`dojo create -name tomsApp` will create a folder named `tomsApp` and scaffold the new app inside.
## Output

screenshot 2016-09-29 16 35 29

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,9 +1,9 @@
...
 <!-- TODO: change and uncomment
-[![Build Status](https://travis-ci.org/dojo/<< package-name >>.svg?branch=master)](https://travis-ci.org/dojo/<< package-name >>)
-[![codecov](https://codecov.io/gh/dojo/<< package-name >>/branch/master/graph/badge.svg)](https://codecov.io/gh/dojo/<< package-name >>)
-[![npm version](https://badge.fury.io/js/dojo-<< package-name >>.svg)](http://badge.fury.io/js/dojo-<< package-name >>)
+[![Build Status](https://travis-ci.org/dojo/cli-create-command.svg?branch=master)](https://travis-ci.org/dojo/cli-create-command)
+[![codecov](https://codecov.io/gh/dojo/cli-create-command/branch/master/graph/badge.svg)](https://codecov.io/gh/dojo/cli-create-command)
+[![npm version](https://badge.fury.io/js/dojo-cli-create-command.svg)](http://badge.fury.io/js/dojo-cli-create-command)
 -->
 
 TODO: Replace with a description of this package
```
Comment (agubler) [inline README.md]: let's add this :)

File: package.json [hunk 2]
```diff
@@ -16,7 +16,7 @@
   ],
   "repository": {
     "type": "git",
-    "url": "https://github.com/tomdye/dojo-cli-create-command.git"
+    "url": "https://github.com/dojo/cli-create-app.git"
   },
   "scripts": {
     "prepublish": "grunt peerDepInstall",
```
Comment (agubler) [inline package.json]: are there any peer deps?

File: src/config/createFilesConfig.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+[
+	[ "package.json", "." ],
+	[ "typings.json", "." ],
+	[ "tsconfig.json", "." ],
+	[ "tslint.json", "." ],
```
Comment (agubler) [inline src/config/createFilesConfig.json]: @matt-gadd questioned whether we want to distribute a `tslint.json` file as part of the app template? I question applied for the `.editorconfig`

File: src/main.ts [hunk 4]
```diff
@@ -1,12 +1,68 @@
...
+
+const packagePath = pkgDir.sync(__dirname);
 
 const command: Command = {
 	description: 'Scaffolds a new command',
-	register(helper) {
+	register(helper: Helper) {
+		helper.yargs.option('n', {
+			alias: 'name',
+			describe: 'The App Name to create',
```
Comment (agubler) [inline src/main.ts]: The name of the application? just small the `The App Name to create` just reads funny to me.

File: src/templates/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+		"noImplicitAny": true,
+		"noImplicitThis": true,
+		"strictNullChecks": true,
+		"outDir": "_build/",
+		"baseUrl": "node_modules",
+		"paths": {},
+		"removeComments": false,
+		"sourceMap": true,
+		"target": "es5",
+		"moduleResolution": "classic"
```
Comment (agubler) [inline src/templates/tsconfig.json]: can we use node resolution?

File: src/templates/package.json [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    "maquette": "^2.3.5"
+  },
+  "devDependencies": {
+    "intern": "^3.2.3",
+    "istanbul": "^0.4.3",
+    "remap-istanbul": "^0.6.4",
+    "tslint": "^3.11.0",
+    "typescript": "2.0.2"
+  }
+}
```
Comment (agubler) [inline src/templates/package.json]: new line

File: src/templates/package.json [hunk 7]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "<%- appName %>",
+  "version": "0.0.0-pre",
+  "description": "A newly scaffolded Dojo 2 App",
+  "private": true,
+  "homepage": "http://<<Your homepage>>",
+  "license": "BSD-3-Clause",
+  "files": [
```
Comment (agubler) [inline src/templates/package.json]: do we want to have the files array?

File: src/templates/package.json [hunk 8]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "<%- appName %>",
+  "version": "0.0.0-pre",
+  "description": "A newly scaffolded Dojo 2 App",
+  "private": true,
```
Comment (agubler) [inline src/templates/package.json]: does it make sense for our d2 apps to have private set to `true` in the package.json

File: src/templates/package.json [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  "homepage": "http://<<Your homepage>>",
+  "license": "BSD-3-Clause",
```
Comment (agubler) [inline src/templates/package.json]: Also should we assume a license?

File: src/templates/typings.json [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+	"globalDevDependencies": {
+		"digdug": "github:dojo/typings/custom/digdug/digdug.d.ts#a62873258aa1deed48f9882c193c335436100d4b",
+		"dojo2": "github:dojo/typings/custom/dojo2/dojo.d.ts#a62873258aa1deed48f9882c193c335436100d4b",
+		"dojo2-dev": "github:dojo/typings/custom/dev/dev.d.ts#288d3a9868194f0e1ad6687371dffd1d96cb39a1",
+		"gruntjs": "registry:dt/gruntjs#0.4.0+20160316171810",
+		"intern": "github:dojo/typings/custom/intern/intern.d.ts#92fcf688c0982c1107fca278de52e4bfe23bd8af",
+		"leadfoot": "github:dojo/typings/custom/leadfoot/leadfoot.d.ts#8f3dfe6cafd04dfa78068785551552a5cfd88db1",
+		"node": "github:dojo/typings/custom/node/node.d.ts#a62873258aa1deed48f9882c193c335436100d4b"
+	}
+}
```
Comment (agubler) [inline src/templates/typings.json]: new line

File: src/typingsInstall.ts [hunk 11]
```diff
@@ -0,0 +1,12 @@
+const spawn: any = require('cross-spawn');
+
+export default async function () {
+	return new Promise((resolve, reject) => {
+		spawn('typings', ['install'], { stdio: 'inherit' })
```
Comment (agubler) [inline src/typingsInstall.ts]: is there not a programmatic API for typings?

File: src/config/createFilesConfig.json [hunk 12]
```diff
@@ -0,0 +1,10 @@
+[
```
Comment (agubler) [inline src/config/createFilesConfig.json]: If you made this a `.ts` file you could do the `path` `join` in here and export it? might make the less verbose `renderFiles`

File: src/createDir.ts [hunk 13]
```diff
@@ -0,0 +1,13 @@
+import { existsSync, mkdirsSync } from 'fs-extra';
+import * as chalk from 'chalk';
+
+export default function (...dirPath: string[]) {
+	dirPath.forEach((path) => {
```
Comment (agubler) [inline src/createDir.ts:5]: do you have to check if the directory exists for each one? couldn't you do this once for the parent directory?

File: README.md (hunk 1 above)
Comment (agubler) [reply README.md]: I appreciate you were probably going to do this as it's still WIP :)

File: package.json (hunk 2 above)
Comment (tomdye) [reply package.json]: Nopes, will remove

File: src/config/createFilesConfig.json (hunk 12 above)
Comment (tomdye) [reply src/config/createFilesConfig.json]: agreed

File: src/config/createFilesConfig.json (hunk 3 above)
Comment (tomdye) [reply src/config/createFilesConfig.json]: What do we think then? Should `tslint` and `editorconfig` be there? I'll raise an issue for discussion

File: src/templates/package.json [hunk 14]
```diff
@@ -0,0 +1,40 @@
+{
```
Comment (matt-gadd) [inline src/templates/package.json:1]: We should drop anything from this file that we cannot determine or adds no value. I think those are: `description`, `homepage`, `license`, `files` and `repository`.  Given the `package.json` lives in the project the user can always add those later. We should probably specify the `version` to be `1.0.0` to align with `npm init` also.

File: src/templates/tsconfig.json (hunk 5 above)
Comment (tomdye) [reply src/templates/tsconfig.json]: is that what we `want` to do?

File: src/typingsInstall.ts (hunk 11 above)
Comment (tomdye) [reply src/typingsInstall.ts]: Is it worth pulling something else in when this works?

File: src/createDir.ts (hunk 13 above)
Comment (tomdye) [reply src/createDir.ts:5]: fair

File: src/templates/tsconfig.json (hunk 5 above)
Comment (agubler) [reply src/templates/tsconfig.json]: yes

File: src/typingsInstall.ts (hunk 11 above)
Comment (agubler) [reply src/typingsInstall.ts]: I would always favour using a programmatic API from a tool like this if one exists (personally)

Comment (tomdye) [reply src/typingsInstall.ts]: i'll take a look then, just think it adds extra size to the command when we're suggesting that they should have it installed globally

File: src/templates/package.json (hunk 14 above)
Comment (tomdye) [reply src/templates/package.json:1]: ok, will do

Comment (tomdye) [reply src/templates/package.json:1]: done

File: src/templates/package.json [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+    "dojo-stores": "^2.0.0-alpha.1",
+    "dojo-widgets": "^2.0.0-alpha.8",
+    "immutable": "^3.8.1",
+    "maquette": "^2.3.5"
+  },
+  "devDependencies": {
+    "intern": "^3.2.3",
+    "istanbul": "^0.4.3",
+    "remap-istanbul": "^0.6.4",
+    "tslint": "^3.11.0",
```
Comment (agubler) [inline src/templates/package.json]: 1) do we need tslint? 2) I don't think this is actually the right version for TS2.0, but we may need to wait until TS2 is released properly for the proper release of the tslint version (currently behind `next` tag)

File: src/templates/tsconfig.json [hunk 16]
```diff
@@ -0,0 +1,22 @@
+{
+	"version": "2.0.2",
+	"compilerOptions": {
+		"declaration": false,
+		"module": "umd",
+		"noImplicitAny": true,
+		"noImplicitThis": true,
+		"strictNullChecks": true,
+		"outDir": "_build/",
+		"baseUrl": "node_modules",
```
Comment (matt-gadd) [inline src/templates/tsconfig.json]: you shouldn't need this or `paths` with `node` resolution

File: src/templates/package.json (hunk 15 above)
Comment (tomdye) [reply src/templates/package.json]: what should we do about it then?

Comment (agubler) [reply src/templates/package.json]: Just remove the dependency is my vote

File: tslint.json [hunk 17]
```diff
@@ -66,4 +66,4 @@
 		"variable-name": [ true, "check-format", "allow-leading-underscore", "ban-keywords" ],
 		"whitespace": [ true, "check-branch", "check-decl", "check-operator", "check-module", "check-separator", "check-type", "check-typecast" ]
 	}
-}
+}
```
Comment (agubler) [inline tslint.json]: new line buddy

**Comment by tomdye** [comment]
All of @agubler comments are addressed.

File: src/npmInstall.ts [hunk 18]
```diff
@@ -0,0 +1,11 @@
+const cs: any = require('cross-spawn');
+
+export default async function () {
+	return new Promise((resolve, reject) => {
+		cs.spawn('npm', ['install'], { stdio: 'inherit' })
```
Comment (novemberborn) [inline src/npmInstall.ts]: https://www.npmjs.com/package/execa may be useful.

File: src/renderFiles.ts [hunk 19]
```diff
@@ -0,0 +1,12 @@
+import template from './template';
+
+export type RenderFilesConfig = {
+	src: string;
+	dest: string;
+}[];
+
+export default async function (renderFilesConfig: RenderFilesConfig, renderData: any) {
+	await renderFilesConfig.forEach(({ src, dest }) => {
```
Comment (novemberborn) [inline src/renderFiles.ts]: `forEach` doesn't return a promise, so why the async / await?

File: src/run.ts [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+
+	if (existsSync(appName)) {
+		return Promise.reject(new Error('App directory already exists'));
+	}
+
+	try {
+		console.info(chalk.underline('Creating Directories'));
+		createDir(...getDirectoryNames(appName));
+
+		changeDir(appName);
```
Comment (novemberborn) [inline src/run.ts]: This feels dirty, explicitly setting `cwd` arguments would be cleaner.

File: src/run.ts [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+		await renderFiles(getRenderFilesConfig(packagePath), { appName });
+
+		console.info(chalk.underline('\n Running npm install'));
+		await npmInstall();
+
+		console.info(chalk.underline('\n Running typings install'));
+		await typings.install({ cwd: '.'});
+	}
+	catch (error) {
+		return Promise.reject(error);
```
Comment (novemberborn) [inline src/run.ts]: Given that you're in an async function, this try/catch is unnecessary. It'll always return a promise.

File: src/run.ts [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	console.info(chalk.green.bold('\nAll done!\n'));
+
+	return Promise.resolve();
```
Comment (novemberborn) [inline src/run.ts]: Unnecessary, the async function will always return a promise.

File: tests/unit/changeDir.ts [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+import * as assert from 'intern/chai!assert';
+import changeDir from 'src/changeDir';
+import { stub, SinonStub } from 'sinon';
+
+let changeDirStub: SinonStub;
+
+registerSuite({
+	name: 'changeDir',
+	'setup'() {
+		changeDirStub = stub(process, 'chdir');
```
Comment (novemberborn) [inline tests/unit/changeDir.ts:11]: Another nice trick is to write a module which exports the `process` global, and then mock that module import.

File: src/run.ts (hunk 20 above)
Comment (tomdye) [reply src/run.ts]: It feels like a better approach than setting `cwd` for every function that comes after. The idea is to create the app's directory and then move into it to perform everything else.

File: src/renderFiles.ts (hunk 19 above)
Comment (tomdye) [reply src/renderFiles.ts]: oops, hangover from previous code.

File: src/run.ts (hunk 21 above)
Comment (tomdye) [reply src/run.ts]: good to know, will see if my tests still run

File: tests/unit/changeDir.ts (hunk 23 above)
Comment (tomdye) [reply tests/unit/changeDir.ts:11]: this is already the module that exports the process functionality 😢

Comment (novemberborn) [reply tests/unit/changeDir.ts:11]: I meant:

`process.ts`:

``` ts
export default process;
```

And then `changeDir.ts`:

``` ts
import process from './process';
```

That way you can stub the `./process` import without having to stub the actual `process` object.

File: README.md [hunk 24]
```diff
@@ -29,24 +30,10 @@ Test cases MUST be written using [Intern](https://theintern.github.io) using the
...
-`grunt test:local`
-
-To test against BrowserStack or Sauce Labs run:
-
-`grunt test:browserstack`
-
-or
-
-`grunt test:saucelabs`
+`grunt test:node`
```
Comment (agubler) [inline README.md]: probably best to override the `grunt test` task to run everything it needs for tests

File: package.json [hunk 25]
```diff
@@ -38,10 +38,19 @@
     "grunt-typings": ">=0.1.5",
     "intern": "^3.2.3",
     "istanbul": "^0.4.3",
+    "mockery": "^1.7.0",
     "remap-istanbul": "^0.6.4",
     "sinon": "^1.17.5",
     "tslint": "^3.11.0",
     "typescript": "2.0.2",
```
Comment (agubler) [inline package.json]: make it typescript `^2.0.3`

File: src/templates/package.json [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+    "dojo-stores": "^2.0.0-alpha.1",
+    "dojo-widgets": "^2.0.0-alpha.8",
+    "immutable": "^3.8.1",
+    "maquette": "^2.3.5"
+  },
+  "devDependencies": {
+    "intern": "^3.2.3",
+    "istanbul": "^0.4.3",
+    "remap-istanbul": "^0.6.4",
+    "typescript": "2.0.2"
```
Comment (agubler) [inline src/templates/package.json]: make this typescript `^2.0.3` also

File: src/templates/tsconfig.json [hunk 27]
```diff
@@ -0,0 +1,20 @@
+{
+	"version": "2.0.2",
```
Comment (agubler) [inline src/templates/tsconfig.json]: just a reminder, keep this inline when you update the ts version :)

File: tests/support/template.ejs [hunk 28]
```diff
@@ -0,0 +1 @@
+<%= value %>
```
Comment (agubler) [inline tests/support/template.ejs]: github tells me there is no new line at the end of the file.

File: tests/unit/changeDir.ts (hunk 23 above)
Comment (agubler) [reply tests/unit/changeDir.ts:11]: You could probably have a single module that exports all the things that you need to stub (like `__dirname`)

File: tests/unit/run.ts [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+
+const name = 'testAppName';
+const args = { name };
+const dirNames = [ name, name + '/tests' ];
+const existsSyncStub: SinonStub = stub();
+const createDirStub: SinonStub = stub();
+const renderFilesStub: SinonStub = stub().returns(Promise.resolve());
+const npmInstallStub: SinonStub = stub().returns(Promise.resolve());
+const changeDirStub: SinonStub = stub();
+let pkgDirStub: SinonStub = stub().returns(name);
```
Comment (agubler) [inline tests/unit/run.ts]: why `let`?

File: tests/unit/template.ts [hunk 30]
```diff
@@ -0,0 +1,40 @@
...
+	},
+	'beforeEach'() {
+		writeFileStub = stub(fs, 'writeFile').callsArg(2);
+		mkdirsStub = stub(fs, 'mkdirsSync');
+	},
+	'afterEach'() {
+		writeFileStub.restore();
+		mkdirsStub.restore();
+	},
+	'can render ejs file'() {
```
Comment (agubler) [inline tests/unit/template.ts]: why not use the `async` pattern used else where?

File: README.md (hunk 24 above)
Comment (tomdye) [reply README.md]: It should be already via the `intern.ts` file. So `grunt test` should just work

File: tests/unit/run.ts (hunk 29 above)
Comment (tomdye) [reply tests/unit/run.ts]: used to be changed further down

File: tests/unit/template.ts (hunk 30 above)
Comment (tomdye) [reply tests/unit/template.ts]: because it was copied from somewhere else (old cli implementation) and i missed it. changing

**Comment by tomdye** [comment]
@agubler @novemberborn comments addressed

**Review by agubler — APPROVED** [review]

**Pull request merged**